### PR TITLE
Tokenizer - better docs and validation

### PR DESCRIPTION
**Pull request by keradus**

### CODE REVIEW

File: src/Tokenizer/Token.php [hunk 1]
```diff
@@ -217,6 +222,8 @@ public function getPrototype()
     /**
      * Get token's content.
      *
+     * It shall be used only for getting the content of token, not for checking it against excepted value.
```
Comment (GrahamCampbell) [inline src/Tokenizer/Token.php]: Very long line. Maybe wrap at 80 chars?

Comment (keradus) [inline src/Tokenizer/Token.php]: big :-1: for 80chars, we do not live in 800x600 screens age

Comment (GrahamCampbell) [inline src/Tokenizer/Token.php]: We do when it comes to reading github diffs...

image

Comment (keradus) [inline src/Tokenizer/Token.php]: poor for github.
we don't follow 80chars in code at all

File: src/Tokenizer/Token.php [hunk 2]
```diff
@@ -61,9 +61,14 @@ public function __construct($token)
             $this->isArray = true;
             $this->id = $token[0];
             $this->content = $token[1];
-        } else {
+        } elseif (is_string($token)) {
             $this->isArray = false;
             $this->content = $token;
+        } else {
+            throw new \InvalidArgumentException(sprintf(
```
Comment (SpacePossum) [inline src/Tokenizer/Token.php]: please add utest for this

Comment (keradus) [inline src/Tokenizer/Token.php]: that's what happen when you extract not-related, side-off changes from another, big PR...

Comment (SpacePossum) [inline src/Tokenizer/Token.php]: the downside of nice reviewable PR's and clean commit history ;)

Comment (keradus) [inline src/Tokenizer/Token.php]: tests added

Comment (SpacePossum) [inline src/Tokenizer/Token.php]: :+1: thanks :)

File: tests/Tokenizer/TokenTest.php [hunk 3]
```diff
@@ -285,24 +285,34 @@ public function provideIsWhitespaceCases()
         );
     }
 
-    public function testPropertiesOfArrayToken()
+    /**
+     * @param array|string|mixed $prototype
```
Comment (julienfalque) [inline tests/Tokenizer/TokenTest.php]: I would replace with `mixed` only.

Comment (keradus) [inline tests/Tokenizer/TokenTest.php]: I was wondering how to nicely write it, `array|string` is valid subset of types, while `mixed` is unvalid subset. together it's `array|string|mixed`.

Comment (julienfalque) [inline tests/Tokenizer/TokenTest.php]: `array|string|mixed` sounds like _"can be anything, or an array or a string"_ which feels weird and is still an invalid subset.

Comment (keradus) [inline tests/Tokenizer/TokenTest.php]: done

**Pull request closed without merging**